compiler.py

```diff
@@ -0,0 +1,637 @@
+            it = REG_IT_FMT.format(call.lineno, call.col_offset)
+            start, end, step = 0, '@links', 1
+            inject.append(f'getlink {target.id} {it}')
+        elif isinstance(call.func, ast.Name) and call.func.id == 'range':
+            it = target.id
+            argv = call.args
+            argc = len(argv)
+            if argc == 1:
+                start, end, step = 0, self.as_value(argv[0]), 1
+            elif argc == 2:
+                start, end, step = *map(self.as_value, argv), 1
+            elif argc == 3:
+                start, end, step = map(self.as_value, argv)
+            else:
+                raise CompilerError(ERR_BAD_ITER_ARGS, node)
+        else:
+            raise CompilerError(ERR_UNSUPPORTED_ITER, node)
+
+        self._ins.append(f'set {it} {start}')
+
+        self._ins.append(f'jump {{}} greaterThanEq {it} {end}')
+        # TODO  if the step in range(start,stop,step) is negative, the relevant test is lessThanEq
+        condition = len(self._ins) - 1
+
+        self._ins.extend(inject)
+        for subnode in node.body:
+            self.visit(subnode)
+
+        self._ins.append(f'op add {it} {it} {step}')
+        self._ins.append(f'jump {condition} always')
+        self._ins[condition] = self._ins[condition].format(len(self._ins))
+
+    def visit_FunctionDef(self, node):
+        """This will be called for any* function definition of the form def fname( args ): body """
+
+        # TODO forbid recursion (or implement it by storing and restoring everything from stack)
+        #      Telos: There's no need to actively forbid recursion, though it may be good to warn about dangers of it.
+        #             So long as there is no stack to overflow, recursion can work perfectly fine as a form of infinite
+        #             loop, or the recursion can be escaped with an end() command, a jump_to( label ) command, or simply
+        #             by entering into some other infinite loop and never return-ing. The key point is just to warn that
+        #             recursive calls will overwrite the value of the return pointer (if we follow my suggestion of
+        #             implementing that in a dummy variable to forego needing a stack) and thereby lose track of
+        #             whatever it was that called the function in the first place.  If you instead push return addresses
+        #             onto a stack, and wait to pop them at return-time, then recursion will still be fine, and will
+        #             even return all the way back to the first caller, with the only caveat being that all variables
+        #             are global, effectively including function arguments, so you have to be a bit cautious how
+        #             you write the code.
+        # TODO local variable namespace per-function
+        #      Telos:  I doubt that mindustry's processor-speed is fast enough to really make this worthwhile.

```

Comment:
This isn't about speed. It's about making the following work:

```python
def f():
    for i in range(100):
        print(i)

i = 0
while i != 5:
    f()  # oops `i` is now 99 and now we're inside an infinite loop
    i += 1
```

Reply:
Yep.  My "solution" to that problem is just to warn users that all variables are global, so if you use `i` in two different places like this, you've gotta expect that the two uses will interact with each other, and if you don't want them to interact, then you shouldn't call them both `i`.  

An alternative semi-solution could be to name-mangle all variables in functions to ensure that they will be unique, but that would mess up a reasonable to desire to be able to use some global variables within functions (esp when passing variables to function occupies scarce processor-cycles!).   Or we could set up some scheme for determining which variables should be global and which should be local/mangled, but I don't think the benefits of this would really repay the costs of setting it up.

It's possible that in-line macros could also be a convenient way to help allay some of these worries, and they'd be much more time-efficient than functions, so I'd view that as a higher priority.

Reply:
Name mangling should be the default, but the [`global`](https://docs.python.org/3/reference/simple_stmts.html#the-global-statement) keyword can be used to instruct the compiler it shouldn't actually mangle the name:

```python
x = 10
def func():
    global x
    y = 20  # mangled by func
    x = 20  # not mangled, refers to global
```